Xcode 14 support

janosvitok opened the pull request:
Changelog: Feature: Support for Xcode 14 and new iOS, watchOS, tvOS and macOS versions.
Docs: omit

Add Xcode 14 support

Also add iOS, macOS, tvOs, watchOs versions



Sources:

    apple-clang version: installed Xcode
    SDK versions: https://xcodereleases.com/
    iOS versions: https://en.wikipedia.org/wiki/IOS_version_history
    tvOS versions: https://en.wikipedia.org/wiki/TvOS
    watchOS versions: https://en.wikipedia.org/wiki/WatchOS

I didn't add all the X.Y.Z versions (like 15.3.1), just the X.Y ones.

Closes https://github.com/conan-io/conan/issues/12063

janosvitok committed:
Add Xcode 14 support

- also add iOS, macOS, tvOs, watchOs versions

H-G-Hristov commented on conans/client/conf/__init__.py:
```diff
@@ -124,7 +124,7 @@
-            version: ["5.0", "5.1", "6.0", "6.1", "7.0", "7.3", "8.0", "8.1", "9.0", "9.1", "10.0", "11.0", "12.0", "13", "13.0", "13.1"]
+            version: ["5.0", "5.1", "6.0", "6.1", "7.0", "7.3", "8.0", "8.1", "9.0", "9.1", "10.0", "11.0", "12.0", "13", "13.0", "13.1", "14.0"]
             libcxx: [libstdc++, libc++]
             cppstd: [None, 98, gnu98, 11, gnu11, 14, gnu14, 17, gnu17, 20, gnu20]
```
What about 23, gnu23 ?
https://developer.apple.com/documentation/xcode-release-notes/xcode-14-release-notes

C++23 features are available in Apple Clang 13.1+

H-G-Hristov commented on conans/client/migrations_settings.py:
```diff
@@ -3874,4 +3874,155 @@
+    apple-clang: &apple_clang
+        version: ["5.0", "5.1", "6.0", "6.1", "7.0", "7.3", "8.0", "8.1", "9.0", "9.1", "10.0", "11.0", "12.0", "13", "13.0", "13.1", "14.0"]
+        libcxx: [libstdc++, libc++]
+        cppstd: [None, 98, gnu98, 11, gnu11, 14, gnu14, 17, gnu17, 20, gnu20]
```
As above 
```suggestion
        cppstd: [None, 98, gnu98, 11, gnu11, 14, gnu14, 17, gnu17, 20, gnu20, 23, gnu23]
```

janosvitok replied:
See above :)

H-G-Hristov commented:
@janosvitok What about C++23 support in Apple Clang 13.1+

janosvitok committed:
Apple clang: describe support for C++20/C++23 (conan 2.x version)

Based on https://github.com/Kitware/CMake/blob/master/Modules/Compiler/AppleClang-CXX.cmake @ 4e3e15f9
and own research on Xcode 14.0 RC (build 14A309).

janosvitok commented:
I've collected changes regarding C++ support from https://github.com/Kitware/CMake/blob/master/Modules/Compiler/AppleClang-CXX.cmake and by running some test code.

Note: while clang man page of Xcode 14 states that the default standard is c++14, __cplusplus in actual code is still defined as 199711L.

memsharded reviewed (changes requested):
Thanks very much for this contribution.

This PR is not updating the correct place, which is ``conan/tools/_compilers.py`` for the new 2.0 ready tools. In some cases we will allow backporting to the legacy build system integrations, but it is necessary to do the changes in the new ones first.

janosvitok commented:
I've removed changes in conans/client/build/ and created similar in conan/tools/_compilers.py
I've left the changed test (I'm not sure how it works and I have not setup proper dev env yet). I've changed also test_cppstd.py.

czoido commented on conans/test/unittests/client/build/cpp_std_flags_test.py:
```diff
@@ -202,6 +202,20 @@ def test_apple_clang_cppstd_flags(self):
         self.assertEqual(_make_cppstd_flag("apple-clang", "12.0", "17"), "-std=c++17")
         self.assertEqual(_make_cppstd_flag("apple-clang", "12.0", "20"), "-std=c++2a")
 
+        self.assertEqual(_make_cppstd_flag("apple-clang", "13.0", "17"), "-std=c++17")
+        self.assertEqual(_make_cppstd_flag("apple-clang", "13.0", "gnu17"), "-std=gnu++17")
+        self.assertEqual(_make_cppstd_flag("apple-clang", "13.0", "20"), "-std=c++20")
+        self.assertEqual(_make_cppstd_flag("apple-clang", "13.0", "gnu20"), "-std=gnu++20")
+        self.assertEqual(_make_cppstd_flag("apple-clang", "13.0", "23"), "-std=c++2b")
+        self.assertEqual(_make_cppstd_flag("apple-clang", "13.0", "gnu23"), "-std=gnu++2b")
+
+        self.assertEqual(_make_cppstd_flag("apple-clang", "14.0", "17"), "-std=c++17")
+        self.assertEqual(_make_cppstd_flag("apple-clang", "14.0", "gnu17"), "-std=gnu++17")
+        self.assertEqual(_make_cppstd_flag("apple-clang", "14.0", "20"), "-std=c++20")
+        self.assertEqual(_make_cppstd_flag("apple-clang", "14.0", "gnu20"), "-std=gnu++20")
+        self.assertEqual(_make_cppstd_flag("apple-clang", "14.0", "23"), "-std=c++2b")
+        self.assertEqual(_make_cppstd_flag("apple-clang", "14.0", "gnu23"), "-std=gnu++2b")
+
```
```suggestion
```

This should be removed and tested with the new conan.tools

czoido commented on conans/test/unittests/client/build/cpp_std_flags_test.py:
```diff
@@ -214,6 +228,8 @@ def test_apple_clang_cppstd_defaults(self):
         self.assertEqual(_make_cppstd_default("apple-clang", "10"), "gnu98")
         self.assertEqual(_make_cppstd_default("apple-clang", "11"), "gnu98")
         self.assertEqual(_make_cppstd_default("apple-clang", "12"), "gnu98")
+        self.assertEqual(_make_cppstd_default("apple-clang", "13"), "gnu98")
+        self.assertEqual(_make_cppstd_default("apple-clang", "14"), "gnu98")
```
```suggestion
```

janosvitok replied:
Same as above. I've put it in a separate commit ("v1.x")

czoido commented on conans/test/unittests/tools/build/test_cppstd.py:
```diff
@@ -67,6 +67,8 @@ def test_supported_cppstd_gcc(compiler, compiler_version, values):
     ("apple-clang", "9.5", ['98', 'gnu98', '11', 'gnu11', "14", "gnu14", "17", "gnu17"]),
     ("apple-clang", "10", ['98', 'gnu98', '11', 'gnu11', "14", "gnu14", "17", "gnu17", "20",
                            "gnu20"]),
+    ("apple-clang", "13", ['98', 'gnu98', '11', 'gnu11', "14", "gnu14", "17", "gnu17", "20",
+                           "gnu20", "23", "gnu23"]),
```
```suggestion
```

czoido commented on conans/test/unittests/tools/build/test_cppstd.py:
```diff
@@ -67,6 +67,8 @@ def test_supported_cppstd_gcc(compiler, compiler_version, values):
     ("apple-clang", "9.5", ['98', 'gnu98', '11', 'gnu11', "14", "gnu14", "17", "gnu17"]),
     ("apple-clang", "10", ['98', 'gnu98', '11', 'gnu11', "14", "gnu14", "17", "gnu17", "20",
                            "gnu20"]),
+    ("apple-clang", "13", ['98', 'gnu98', '11', 'gnu11', "14", "gnu14", "17", "gnu17", "20",
+                           "gnu20", "23", "gnu23"]),
```
```suggestion
    ("apple-clang", "13", ['98', 'gnu98', '11', 'gnu11', "14", "gnu14", "17", "gnu17", "20", "gnu20"]),
```

H-G-Hristov replied:
This looks like it will fix the test but is this the real issue?

czoido replied:
but is c++23 supported by apple-clang 13? I could not see mentions in the man page

janosvitok replied:
I will check xcodes for actual support.

janosvitok replied:
clang++ 13.0 accepts c++2b and gnu++2b. clang from 12.5.1. doesn't.

H-G-Hristov replied:
> but is c++23 supported by apple-clang 13? I could not see mentions in the man page

https://developer.apple.com/documentation/xcode-release-notes/xcode-14-release-notes

> Apple Clang Compiler
New Features
New C++ projects you create in Xcode use C++20 language dialect by default. (93456065)
Several C++20 and C++2b papers have been implemented:
C++20 papers that have been implemented:
P0692R1 - Access checking on specializations
P0388R4 - Permit conversions to arrays of unknown bound
C++2b papers that have been implemented:
P1938R3 - if consteval
P1401R5 - Narrowing contextual conversions to bool
P1949R7 - C++ identifier syntax using Unicode Standard Annex 31
P2360R0 - Extend init statement to allow alias declaration (93898598)

https://developer.apple.com/documentation/xcode-release-notes/xcode-13_3-release-notes

> Apple Clang Compiler
New Features
Several C++20 and C++2b papers have been implemented:
using declaration can now be used for enums, enum classes, and their members.
The compiler can deduce the size_t and ssize_t types based on new literal suffixes (uz, z), for example for (auto i = 0uz; i < vector.size(); ++i) {}.
If a lambda declaration takes no parameters, the open and close parentheses () can now be omitted in all contexts including mutable lambdas.
Duplicate attributes are now allowed by the compiler.
Any trailing whitespace after \ is non-significant.
Concatenations of string-literals with different encoding prefixes are now ill-formed, for example, auto s = L"" U"";. (89022082)

janosvitok replied:
> New C++ projects you create in Xcode use C++20 language dialect by default. (93456065)

This means that Xcode sets some project property to c++20, so it calls clang with that. clang itself without --std= still defaults to gnu++98. I've checked this by comparing output of `clang++ -x c++ -dM -E /dev/null` with output of `clang++ -x c++ -dM --std=c++98 -E /dev/null` and `clang++ -x c++ -dM --std=gnu++98 -E /dev/null`. I've also checked __cplusplus value in a compiled file.

Manual pages are wrong; they are most probably taken from vanilla clang.

I've checked list of supported --std= values by running `clang++ --std=123 test.cpp`:
* clang++ from Xcode 12.5.1 (12.0.5/clang-1205.0.22.11) **does not** support c++2b/gnu++2b
* clang++ from Xcode 13.0 (13.0.0/clang-1300.0.29.3) **does** support c++2b/gnu++2b

H-G-Hristov replied:
Apple Clang supports already a fair amount of C++2b features and I'm looking to enabled them in our projects as soon as we get the Conan support :)

https://en.cppreference.com/w/cpp/23

janosvitok replied:
I've updated the 2.x code so the test passes.

Note that there is duplicate information in `conan/tools/_compilers.py` and `conan/tools/build/cppstd.py`. I think the former could be implemented using the latter (something like check the flag for all known standards, and return all tha have flag != None).

H-G-Hristov commented:
> Thanks very much for this contribution.
> 
> This PR is not updating the correct place, which is `conan/tools/_compilers.py` for the new 2.0 ready tools. In some cases we will allow backporting to the legacy build system integrations, but it is necessary to do the changes in the new ones first.

I hope you'll continue supporting 1.5x.x series for a while especially with the new Xcode stuff. It seems we won't be able to use 2.0 any time soon. We have too many custom recipes and nobody to update them.
I have the impression they won't be compatible with 2.0.

janosvitok committed:
Apple clang: describe support for C++20/C++23 (conan 1.x version)

janosvitok commented:
Guys,

I've sorted out the changes:

1. add support for apple-clang 14, and various apple OS version to conan settings
2. describe C++23 support in apple-clang in conan 2.x flavor (conan/tools/*)
3. describe the same in conan 1.x flavor (conans/client/*)

I guess at this point I've done my best, and anything beyond is up to you :)
Feel free to do anything with the code.

I still will try to do changes if requested.

From my point of view, we're so far on 1.52, and don't need C++23 (we're using conan to pre-compile thirdparty libs),
so I really need only the first commit (so that **any** compilation can be done with Xcode 14).

memsharded commented:
Thanks very much for the contribution @janosvitok, I have assigned it to next release, we will manage it.

janosvitok committed:
Fix test

The test doesn't set compiler version (just OS version), so I suppose the latest is used.

janosvitok commented:
I've added one more commit to fix the failing test. It seems that the test checks the latest compiler.

Note: I've checked the C++2a flag: 12.0 knows only c++2a, 12.5.1 knows c++20 (and accepts c++2a), so it must have changed somewhere between. Nevertheless 13.0 is a safe bet (since 2a is accepted everywhere).

H-G-Hristov commented on conan/tools/_compilers.py:
```diff
@@ -236,24 +236,32 @@ def _cppstd_apple_clang(clang_version, cppstd):
+        vgnu17 = "gnu++1z"
 
-    if Version(clang_version) >= "10.0":
+    if Version(clang_version) >= "13.0":
+        v20 = "c++20"
+        vgnu20 = "gnu++20"
+    elif Version(clang_version) >= "10.0":
         v20 = "c++2a"
         vgnu20 = "gnu++2a"
 
+    if Version(clang_version) >= "13.0":
+        v23 = "c++2b"
+        vgnu23 = "gnu++2b"
```
Just a nit: Why don't you merge the two `if`? It will be more clear/obvious that something changed in `13.0`. If you do. Don't forget to do it below too.

janosvitok replied:
I've followed current structure of the code. There's one section per c++ standard version. Those if are the same, but the latter will be converted to if / elif once "c++23" switch is accepted.

czoido commented on conans/client/conf/__init__.py:
```diff
@@ -124,9 +124,9 @@
             runtime_type: [None, Debug, Release]
         apple-clang: &apple_clang
-            version: ["5.0", "5.1", "6.0", "6.1", "7.0", "7.3", "8.0", "8.1", "9.0", "9.1", "10.0", "11.0", "12.0", "13", "13.0", "13.1"]
+            version: ["5.0", "5.1", "6.0", "6.1", "7.0", "7.3", "8.0", "8.1", "9.0", "9.1", "10.0", "11.0", "12.0", "13", "13.0", "13.1", "14.0"]
```
```suggestion
            version: ["5.0", "5.1", "6.0", "6.1", "7.0", "7.3", "8.0", "8.1", "9.0", "9.1", "10.0", "11.0", "12.0", "13", "13.0", "13.1", "14", "14.0"]
```
As Conan will autodetect for apple-clang version>=13 the version with just the major number, I think that change should be done

janosvitok replied:
I agree.

janosvitok replied:
I've added "14" in the last commit.

H-G-Hristov commented on conans/client/conf/__init__.py:
```diff
@@ -40,28 +40,28 @@
                       "13.0", "13.1", "13.2", "13.3", "13.4", "13.5", "13.6", "13.7",
                       "14.0", "14.1", "14.2", "14.3", "14.4", "14.5", "14.6", "14.7", "14.8",
-                      "15.0", "15.1", "15.2", "15.3", "15.4"]
+                      "15.0", "15.1", "15.2", "15.3", "15.4", "15.5", "15.6", "16.0"]
```
There is already Xcode 14.1 Beta in the works which is targeting iOS 16.1, TV OS 16.1. Should these be added now?
https://developer.apple.com/documentation/xcode-release-notes/xcode-14_1-release-notes
https://xcodereleases.com

janosvitok replied:
Yes. I am downloading new xcodes an will report what version do they report.

janosvitok replied:
14.1 beta:

> Apple clang version 14.0.0 (clang-1400.0.29.201)
> Target: x86_64-apple-darwin21.6.0

janosvitok replied:
I've added commit with new iOS/watchOS/tvOS versions. Do we want to add all macOS minor versions?
There are too many of them...

janosvitok committed:
Add new Apple OS versions.
